lib/livebook_web/live/session_live/secrets_component.ex

```diff
@@ -164,23 +270,71 @@ defmodule LivebookWeb.SessionLive.SecretsComponent do
   defp prefill_secret_name(socket) do
     case socket.assigns.prefill_secret_name do
       nil ->
-        if unavailable_secret?(socket.assigns.preselect_name, socket.assigns.secrets),
-          do: socket.assigns.preselect_name,
-          else: ""
+        if unavailable_secret?(
+             socket.assigns.preselect_name,
+             socket.assigns.secrets,
+             socket.assigns.livebook_secrets
+           ),
+           do: socket.assigns.preselect_name,
+           else: ""
 
       prefill ->
         prefill
     end
   end
 
-  defp unavailable_secret?(nil, _), do: false
-  defp unavailable_secret?("", _), do: false
+  defp unavailable_secret?(nil, _, _), do: false
+  defp unavailable_secret?("", _, _), do: false
 
-  defp unavailable_secret?(preselect_name, secrets) do
-    preselect_name not in Enum.map(secrets, & &1.name)
+  defp unavailable_secret?(preselect_name, secrets, livebook_secrets) do
+    preselect_name not in Enum.map(secrets, & &1.name) &&
+      preselect_name not in Enum.map(livebook_secrets, & &1.name)
   end
 
   defp title(%{assigns: %{select_secret_ref: nil}}), do: "Add secret"
   defp title(%{assigns: %{select_secret_options: %{"title" => title}}}), do: title
   defp title(_), do: "Select secret"
+
+  defp put_secret(pid, secret, "session"), do: Livebook.Session.put_secret(pid, secret)
+  defp put_secret(_pid, secret, "livebook"), do: Livebook.Secrets.set_secret(secret)
+
+  defp grant_access(secret_name, socket) do
+    secret_value =
+      Enum.find_value(
+        socket.assigns.livebook_secrets,
+        &if(&1.name == secret_name, do: &1.value)
+      )
+
+    secret = %{name: secret_name, value: secret_value}
+    put_secret(socket.assigns.session.pid, secret, "session")
+  end
+
+  defp livebook_only_secrets(secrets, livebook_secrets) do
+    Enum.reject(livebook_secrets, &(&1.name in get_in(secrets, [Access.all(), :name])))
```

Comment:
You want to compute the names before hand instead of inside each loop:

```suggestion
    secret_names = get_in(secrets, [Access.all(), :name])
    Enum.reject(livebook_secrets, &(&1.name in secret_names))
```